src/debugAdapter/goDebug.ts

```diff
@@ -479,65 +479,63 @@ class Delve {
 		});
 	}
 
-	callPromise<T>(command: string, args: any[]): Thenable<T> {
+	public callPromise<T>(command: string, args: any[]): Thenable<T> {
 		return new Promise<T>((resolve, reject) => {
 			this.connection.then(conn => {
-				conn.call<T>('RPCServer.' + command, args, (err, res) => {
-					if (err) return reject(err);
-					resolve(res);
+				conn.call<T>(`RPCServer.${command}`, args, (err, res) => {
+					return err ? reject(err) : resolve(res);
 				});
 			}, err => {
 				reject(err);
 			});
 		});
 	}
 
-	close(): Thenable<void> {
+	public close(): Thenable<void> {
 		log('HaltRequest');
 
-		return new Promise(resolve => {
-			let timeoutToken: NodeJS.Timer;
-			if (this.debugProcess) {
-				timeoutToken = setTimeout(() => {
-					log('Killing debug process manually as we could not halt and detach delve in time');
-					killTree(this.debugProcess.pid);
-					resolve();
-				}, 1000);
+		return new Promise(async resolve => {
+			const timeoutToken: NodeJS.Timer = this.debugProcess && setTimeout(() => {
+				log('Killing debug process manually as we could not halt and detach delve in time');
+				killTree(this.debugProcess.pid);
+				resolve();
+			}, 1000);
+
+			try {
+				await this.callPromise('Command', [{ name: 'halt' }]);
+			}
+			catch (err) {
+				const errMsg = err ? err.toString() : '';
+				log(`Failed to halt - ${errMsg}`);
+			}
+
+			if (timeoutToken) {
+				clearTimeout(timeoutToken);
 			}
 
-			this.callPromise('Command', [{ name: 'halt' }]).then(() => {
-				if (timeoutToken) {
-					clearTimeout(timeoutToken);
+			log('HaltResponse');
+			if (!this.debugProcess) {
+				log('RestartRequest');
+				try {
+					await this.callPromise('Restart', this.isApiV1 ? [] : [{ position: '', resetArgs: false, newArgs: [] }]);
```

Comment:
@ramya-rao-a a bit unrelated but do you know why we attempt to restart the debuggee during close?  I believe this is related to remote debugging as that's the only time `this.debugProcess` is null.

Reply:
Yes, in case of remote debugging, we don't start the delve process. It was started by the user beforehand. Therefore, we shouldn't be killing the delve process.

I wonder if instead of restart, we should just detach without the kill switch.

@vladbarosan Can you give that a try?

Reply:
i think that might make sense, I think Delve in API mode can accept multiple clients, in that case I dont think we would want to restart it.

Reply:
also, given those, does it even make sense to halt a remote program ?

Reply:
we definitely need to detach. And since we can't detach without halting, we need to halt.

Are you saying that we dont need to detach either?

Reply:
well i think we need to detach, maybe im not familiar enough with the remote debugging and I need to try it . I thought delve is started outside our responsability already with a target program ? then do we really need to halt it from our side in that case ?

Reply:
> I thought delve is started outside our responsability already with a target program

Yes, that's right. Delve is started by the user in the remote machine.

> then do we really need to halt it from our side in that case

I believe we need to detach, else the delve process will continue to think that there is a client attached to it.

Reply:
yes agree to `detach`, was saying we dont need to `halt` if the above mentioned are correct. So do we agree on the path forward here ?
for remote, don't even halt, just detach ? if so I can have a look see if everything goes well in that in this direction.

Reply:
> for remote, don't even halt, just detach

The problem is that we can't detach without halting a delve process (regardless of its remote or not)

Reply:
oh really ? that's a bummer, i would expect delve to allow you to detach without halting. ill have a look and see what we can do for remote.